libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/NavigationRouteEngine.java

```diff
@@ -0,0 +1,195 @@
+package com.mapbox.services.android.navigation.v5.route;
+
+import android.location.Location;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.mapbox.api.directions.v5.models.DirectionsResponse;
+import com.mapbox.api.directions.v5.models.RouteOptions;
+import com.mapbox.core.utils.TextUtils;
+import com.mapbox.geojson.Point;
+import com.mapbox.services.android.navigation.v5.navigation.NavigationRoute;
+import com.mapbox.services.android.navigation.v5.navigation.NavigationUnitType;
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;
+import com.mapbox.services.android.navigation.v5.utils.RouteUtils;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+import timber.log.Timber;
+
+/**
+ * This class can be used to fetch new routes given a {@link Location} origin and
+ * {@link RouteOptions} provided by a {@link RouteProgress}.
+ */
+public class NavigationRouteEngine implements RouteEngine {
+
+  private static final double BEARING_TOLERANCE = 90d;
+
+  private List<RouteEngineListener> routeEngineListeners = new CopyOnWriteArrayList<>();
+
+  private String accessToken;
+  private String routeProfile;
+  private RouteProgress routeProgress;
+  private Locale locale = Locale.getDefault();
+  private int unitType = NavigationUnitType.NONE_SPECIFIED;
+
+  @Override
+  public void addRouteEngineListener(RouteEngineListener listener) {
+    if (!routeEngineListeners.contains(listener)) {
+      routeEngineListeners.add(listener);
+    }
+  }
+
+  @Override
+  public void removeRouteEngineListener(RouteEngineListener listener) {
+    routeEngineListeners.remove(listener);
+  }
+
+  @Override
+  public void updateAccessToken(String accessToken) {
+    this.accessToken = accessToken;
+  }
+
+  @Override
+  public void updateLocale(Locale locale) {
+    this.locale = locale;
+  }
+
+  @Override
+  public void updateUnitType(int unitType) {
+    this.unitType = unitType;
+  }
+
+  @Override
+  public void updateRouteProfile(String routeProfile) {
+    if (RouteUtils.isValidRouteProfile(routeProfile)) {
+      this.routeProfile = routeProfile;
+    }
+  }
+
+  /**
+   * Calculates a new {@link com.mapbox.api.directions.v5.models.DirectionsRoute} given
+   * the current {@link Location} and {@link RouteProgress} along the route.
+   * <p>
+   * Uses {@link RouteOptions#coordinates()} and {@link RouteProgress#remainingWaypoints()}
+   * to determine the amount of remaining waypoints there are along the given route.
+   *
+   * @param location      current location of the device
+   * @param routeProgress for remaining waypoints along the route
+   * @since 0.13.0
+   */
+  @Override
+  public void findRouteFromRouteProgress(Location location, RouteProgress routeProgress) {
```

Comment:
We could refactor `findRouteFromRouteProgress` (decompose conditionals, extract blocks of code into well-named methods, etc.). We should keep `public` methods as clean as possible so they read like pseudocode (i.e. no conditionals, fors, etc.).